ZeldaFullEditor/Gui/MainTabs/TextEditor.cs

```diff
@@ -15,693 +15,966 @@
+			}
+
+			public string GetParameterizedToken(byte value = 0)
+			{
+				if (HasArgument)
+				{
+					return $"[{Token}:{value:X2}]";
+				}
+				else
+				{
+					return $"[{Token}]";
+				}
+			}
+
+			public override string ToString()
+			{
+				return $"{GenericToken} {Description}";
+			}
+
+			public Match MatchMe(string dfrag)
+			{
+				return Regex.Match(dfrag, StrictPattern);
+			}
+		}
+
+		public class ParsedElement
+		{
+			public TextElement ElementType { get; }
+
+			public byte Value { get; }
+
+			public ParsedElement(TextElement textElement, byte value)
+			{
+				ElementType = textElement;
+				Value = value;
+			}
+		}
+
+		public class DictionaryEntry
+		{
+			public byte ID { get; }
+
+			public int RealID => ID + DictionaryBase;
+
+			public string Contents { get; }
+
+			public byte[] Data { get; }
+
+			public int Length { get; }
+
+			public string Token { get; }
+
+			public DictionaryEntry(byte id, string contents)
+			{
+				Contents = contents;
+				ID = id;
+				Length = contents.Length;
+				Token = $"[{DictionaryToken}:{ID:X2}]";
+				Data = ParseMessageToData(Contents);
+			}
+
+			public bool ContainedInString(string str)
+			{
+				return str.IndexOf(Contents) >= 0;
+			}
+
+			public string ReplaceInstancesOfIn(string str)
+			{
+				return str.Replace(Contents, Token);
+			}
+
+			public string ToPrettyString()
+			{
+				return $"{ID:X2} [{ID + DictionaryBase:X2}] - {Contents.Replace(" ", "[Space]")}";
+			}
+		}
+
+		public static string ReplaceAllDictionaryWords(string str)
+		{
+			string temp = str;
+
+			foreach (DictionaryEntry dictionaryEntry in AllDictionaries)
+			{
+				if (dictionaryEntry.ContainedInString(temp))
+				{
+					temp = dictionaryEntry.ReplaceInstancesOfIn(temp);
+				}
+			}
+
+			return temp;
+		}
+
+		/// <summary>
+		///
+		/// </summary>
+		/// <param name="str"> The string to find. </param>
+		/// <returns></returns>
+		public static ParsedElement FindMatchingElement(string str)
+		{
+			Match match;
+
+			foreach (TextElement textElement in TextCommands.Concat(SpecialChars))
+			{
+				match = textElement.MatchMe(str);
+
+				if (match.Success)
+				{
+					if (textElement.HasArgument)
+					{
+						return new ParsedElement(textElement, byte.Parse(match.Groups[1].Value, NumberStyles.HexNumber));
+					}
+					else
+					{
+						return new ParsedElement(textElement, 0);
+					}
+				}
+			}
+
+			// See if dictionary entry.
+			match = DictionaryElement.MatchMe(str);
+			if (match.Success)
+			{
+				return new ParsedElement(
+					DictionaryElement,
+					(byte) (DictionaryBase + byte.Parse(match.Groups[1].Value, NumberStyles.HexNumber)));
+			}
+
+			return null;
+		}
+
+		public TextElement FindMatchingCommand(byte b)
+		{
+			return TextCommands.FirstOrDefault(e => e.ID == b);
+		}
+
+		public TextElement FindMatchingSpecial(byte value)
+		{
+			return SpecialChars.FirstOrDefault(e => e.ID == value);
+		}
+
+		// TODO make this return the actual dictionary entry
```

Comment:
"Modify this routine to take the dictionary as argument and return the value rather than the position."